R/tag_env.R

```diff
@@ -0,0 +1,1308 @@
+#
+# Disadvantages
+# * MUST be careful not alter the environment object before converting back to a list. (Ex: Do not remove the element environment's children)
+# * The item returned is a set of environments that will alter in place. We will need to be careful about documenting and/or safeguarding this
+
+
+# ## Final design choice:
+# Use environment elements
+# * Being able to search and have a list of eles to immediately look at and alter in place is AMAZING!
+# * Being able to ask for a grandparent (or obj$parent$parent) and be able to alter it in place is AMAZING! This has a strongly influenced by jquery.
+
+# ----------
+
+# # Current design decisions
+# * Tag graphs or tag environments can NOT be used in UI. These objects MUST be converted back to standard tag objects.
+# * tagFunctions will not be altered in place
+#   * To alter tagFunction()s, use the `onRender(x)` method to register a method to be called after `as.tags(x)` is called.
+#   * `onRender(x, expr)` will wrap create a tag function that will resolve the tags before running the expr.
+
+
+# Wrap in function so r cmd check is happy
+envir_key_fn <- function(x) { format.default(x) }
+# No need to contantly convert the envir to a key string. The value should be constant given an envir
+envir_key_or_stop <- function(x) {
+  x$env_key %||% stop("`x` is not a tag environment. `x$env_key` is missing")
+}
+
+# use for `has()` functionality
+envir_map <- function() {
+  map <- fastmap()
+  list(
+    keys = function() {
+      map$keys()
+    },
+    as_list = function() {
+      unname(map$as_list())
+    },
+    has = function(envir) {
+      map$has(envir_key_or_stop(envir))
+    },
+    set = function(envir, value = envir) {
+      map$set(envir_key_or_stop(envir), value)
+    },
+    remove = function(envir) {
+      map$remove(envir_key_or_stop(envir))
+    }
+  )
+}
+# use for consistent `as_list()` order
+envir_stack <- function() {
+  stack <- faststack()
+  list(
+    push = stack$push,
+    as_list = stack$as_list,
+    unique_list = function() {
+      unique(stack$as_list())
+    }
+  )
+}
+
+
+
+
+
+# Retrieve all attributes that can be manually set
+# ?attr
+# Note that some attributes (namely ‘class’, ‘comment’, ‘dim’,
+# ‘dimnames’, ‘names’, ‘row.names’ and ‘tsp’) are treated specially
+# and have restrictions on the values which can be set.
+safe_attr_values <- function(x) {
+  bad_el_attrs <- c("class", "comment", "dim", "dimnames", "names", "row.names", "tsp")
+  attr_vals <- attributes(x)
+  attr_vals[bad_el_attrs] <- NULL
+  attr_vals
+}
+
+# Convert a list to an environment and keep class and attribute information
+safe_list_2_env <- function(x, new_class = NULL) {
+  x_list <- x
+
+  ret <- list2env(x_list, new.env(parent = emptyenv()))
+
+  attr_vals <- safe_attr_values(x_list)
+  Map(names(attr_vals), attr_vals, f = function(attr_name, attr_value) {
```

Comment:
Can you use `walk2` from https://github.com/wch/staticimports/blob/main/inst/staticexports/map.R ?

It will be slightly faster.